### ISSUE TITLE
Times (sometimes) change on page load

### ISSUE THREAD

**Issue by ajmueller** [beenote/angular-material-datetimepicker#48]
**Library Versions:**
- AngularJS: 1.6.10
- Angular Material: 1.1.10
- Angular Material Datetimepicker: 1.17.1
- Moment.js: 2.19.3
- MomentTimezone.js: 0.5.13

**Description of the Issue:**
We store the datetime of events in our database alongside the timezone offset in hours (e.g. `-0700`) and the timezone name of the event's location (e.g. `America/Los_Angeles`).  The timezone offset is used with `ng-model-options` on the datetimepicker to correctly set and display the time of the event, even if you aren't in the same timezone as the location of the event.  When loading the edit event page, the initial value of the datetime will *sometimes* shift by an hour either forward or backward.  This doesn't seem to have any tie to daylight savings time changing soon in the United States, either (that happens on November 4th).

For example, we have an event on Friday October 26 at 8am and loading the page has resulted in the time changing on load to 7am.  It also happened with an event on Monday October 29 at 11am when it loaded to show 12pm.

Unfortunately this issue is inconsistent and, from what we've been able to find, unpredictable to reproduce.  Some users don't see it at all while others see it a couple times a day (when editing, on average, a couple dozen or so events per day).  I've also seen it with my own eyes in production, but have been unable to reproduce it locally or in a Plunker.

We did recently switch from version 1.12.0 to 1.17.1 to get the fix for #47; that deployment did not change any other event editing code.  The deployment happened 6 days ago and this new issue did not appear until that time.  As such, it seems possible that a change between those two versions is leading to this.  Looking at the [file changes from 1.12.0 to 1.17.1](https://github.com/beenote/angular-material-datetimepicker/compare/v1.12.0...v1.17.1), there appears to be quite a few changes around the usage of `ng-model-options` that were introduced in the [1.13.0 update](https://github.com/beenote/angular-material-datetimepicker/compare/v1.12.0...v1.13.0).  Since we updated from 1.12.0 directly to 1.17.1, we haven't tested any versions in between to know if they also contain the bug or not.

I wish I had more information and a very clear example to provide you like I have with past bug reports, but we haven't been able to crack it.  We're hoping the above describes the situation well enough that it may trigger a thought in your head about what could cause such an issue.  In the meantime, we will continue to look at this on our end and see if we can pinpoint the exact problem.

Thanks as always for your help and continued support of this library.

**EDIT** for what it's worth, it does appear that this only happens to users with slower computers.  All users that I've spoken to are using Chrome, too.

**Comment by hexadecy** [comment]
Ok I created a special version with the fix #47 only: [version 1.12.1](https://github.com/beenote/angular-material-datetimepicker/releases/tag/v1.12.1). 

- Please test with a slow connection : https://helpdeskgeek.com/networking/simulate-slow-internet-connection-testing
- Test with a slow cpu: https://stackoverflow.com/questions/8521683/is-there-a-way-to-throttle-javascript-performance-to-simulate-a-slow-client.
- Test with Internet Explorer 10-11.
- Do you have the issue on mobile too?

**Comment by ajmueller** [comment]
Awesome, thanks so much @hexadecy!  I'll switch to this version and put it in our dev environment immediately.  Regarding testing:

* We did run tests simulating slow connections and CPUs in Chrome for 1.17.1.  We'll be sure to the same for 1.12.1.  Unfortunately, in both cases (and even when they were combined), we were unable to reproduce the issue on our dev machines.
* I'll make sure to test IE11 as well.  We don't have any users on IE10 or lower.
* I haven't heard of the issue happening on mobile, though our user base that accesses the page where this is happening is almost entirely desktop-browser based.

Thanks again for the quick help and I'll be sure to keep you updated on how this works for us.

**Comment by hexadecy** [comment]
My guess is with 1.13.3, if it's a moment date it will not be converted back to a js date as before. Did you pass a moment or a js string date to the model?

**Comment by ajmueller** [comment]
@hexadecy sorry for the delay, I took yesterday off work.

We do, in fact, pass a Moment object to the model and not a JS date.

**Comment by hexadecy** [comment]
I did a lot of work for timezone in https://github.com/beenote/angular-material-datetimepicker/issues/31. Do you use `edit-input=true`. 

@ajmueller  Can you console log a typical moment object from your backend server, I want to have a look at ._tzm and ._isUTC.

**Comment by ajmueller** [comment]
@hexadecy I ended up figuring out the issue and I feel awful because it's entirely on our end and not in the picker at all.  While putting together a Plunkr example that more holistically represents our usage, it suddenly hit me what was happening and the timing with our upgrade to 1.17.1 just happened to coincide with an increase in managing events during both daylight time and standard time.

Long story short, it's always important to make sure your scoped variable values are what you expect in a service with shared data.

And now I owe you a 🍺 or ☕️ .  Or a week's worth 😄

**Comment by hexadecy** [comment]
@ajmueller I'm glad it's on your side 😄. Happy 🎃!

**Comment by ajmueller** [comment]
I'm glad it's on our side as well.  Happy 🎃 to you, too!
